### ISSUE TITLE
Compression support

### ISSUE THREAD

**Issue by vlmutolo** [cberner/redb#301]
Transparent compression is one of the killer features for both RocksDB and sled. It's especially helpful when you have huge numbers of keys (100M+) and you're storing easily compressible values, like JSON data. 

Are there plans to add it to redb? I'd personally want one or both of lz4(hc) and zstd as options (EDIT: if the maintainers are interested in adding page-level compression).

**Comment by cberner** [comment]
There aren't any immediate plans. Have you found that transparent compression is especially valuable over compressing and then storing the compressed value, on top of the database?

**Comment by vlmutolo** [comment]
Yes, page-level compression can be extremely valuable. The most common use case is probably if you have many small JSON values with the same structure. For example, let's say we're storing some basic info on requests a web service sees, and we're storing it in JSON that looks like this:

```json
{"endpoint": "/user/profile", "user": "userA", "…": "…"}
{"endpoint": "/user/profile", "user": "userB", "…": "…"}
{"endpoint": "/user/signout", "user": "userA", "…": "…"}
```

A single one of these objects will compress relatively poorly compared to compressing a page filled with them. Even a byte-level, cheap compression algorithm like LZ4HC should be able to identify `endpoint` and `user` as repeated keys that should be stored in a dictionary. Plus, repeated values like `/user/profile` would also be compressed.

In general I think there are lots of cases where there's significant duplication *across* keys and values but nothing *within* a single key or value.

**Comment by vlmutolo** [comment]
If there's interest in adding this as a feature, I could help with the implementation. Skimming the code a bit, it seems like pages are currently read directly from a memory map. So I think redb would need to have some kind of uncompressed page cache.

**Comment by cberner** [comment]
Ah yes, I see, that use case makes sense.

I'm not sure about adding this feature, because I think it would significantly complicate the codebase -- a userspace page cache sounds particular annoying because users will then have to tune its size. I haven't looked at how sled & RocksDB implement it though, so maybe there's a simple implementation that I'm not thinking of. Do you know how their implementations are designed?

**Comment by vlmutolo** [comment]
Sled has a userspace page cache instead of using a memory map, which makes it easier for them.

https://github.com/spacejam/sled/blob/e95ec0571ae985e6f763a8226db8b8d6c065daba/src/pagecache/mod.rs#L466-L486

If compression is enabled, sled decompresses the page when reading from disk. This is simpler than redb's use case because there was already a userspace cache, and the only difference when using compression is decompressing the data right after reading it in.

https://github.com/spacejam/sled/blob/e95ec0571ae985e6f763a8226db8b8d6c065daba/src/pagecache/heap.rs#L336

I'm not as good at navigating C++ codebases, but based on [this RocksDB blog post](http://rocksdb.org/blog/2021/05/27/rocksdb-secondary-cache.html) they have a page cache. RocksDB apparently also has a *block* cache so the primary store can be on "remote storage or cloud storage". Pretty cool.

---

I think redb could implement page-level compression while retaining the current mmap-based, OS-controlled cache. I really like the idea of letting the OS handle caching reads. It basically lets the page cache safely grow to available memory.

Redb could have an additional, much smaller userspace cache that holds only the uncompressed data. I took a look through the code base, and I was thinking something like the following might work.

```rust
pub(crate) trait Page {
    fn memory(&self, cage: &PageCache) -> PageData;
    fn get_page_number(&self) -> PageNumber;
}

pub(crate) enum PageData<'a> {
    Cached(Arc<[u8]>),
    Direct(&'a [u8]),
}

impl<'a> AsRef<[u8]> for PageCache<'a> { … }

pub(crate) struct PageCache {
    raw: Mutex<LruCache<PageNumber, PageData<'static>>>,
}
```

Basically, change the `Page` trait to output a type (`PageCache`) that can either be an owned `Arc` or the normal mmap'ed byte slice that it currently uses. `PageCache` would implement `AsRef<[u8]>` so that changes to all the places in the code that *consume* from this type are minimal.

I didn't finish looking into it. We'd need to find a way to pass the `&PageCache` around (though there's already a `PageAllocator` so maybe they should be joined?) and also what to do about writes. I imagine writes are easier because the compressed data could just be written directly to the mmap'ed buffer.

**Comment by cberner** [comment]
The `PageAllocator` is just a data structure that tracks which pages are allocated. The place to add a cache would probably be in `TransactionalMemory`, but I'd be a bit worried that adding a lookup into this cache for every page lookup will hurt performance for workloads that don't use compression. There's a `open_dirty_pages` set that's only enabled in debug mode, because iirc, it reduced performance.

I think there's at least one place where the lifetimes of that approach may not work. Have a look at this line: https://github.com/cberner/redb/blob/5b11d03f9dd473e9534d7ba8e0965310aa8a2b22/src/tree_store/btree.rs#L275
The `Page` object is converted into a slice with the appropriate lifetime, so that it can be zero-copy converted into the type that is returned.

At a higher-level though, can you tell me more about the use case you have that needs compression? Do you need keys to be compressed, or just values? Is the compression needed to improve performance, or just reduce storage?

**Comment by vlmutolo** [comment]
I'll take a closer look at the code soon, but I can comment on the use case. Compression for the keys would be nice, but the values are significantly larger than the keys. The values are around 1kB each and have a lot of redundant structure across values. The compression would be to reduce the storage requirements. 

---

As an interesting aside, I just saw today that someone [wrote an extension](https://phiresky.github.io/blog/2022/sqlite-zstd/) (in Rust) for SQLite to add zstd page compression. So I'll probably be adding that to my list of key-value stores to try.

**Comment by cberner** [comment]
Cool, ya that seems like a good option to try! Another thing you could try is compression at the file system level. btrfs supports transparent compression and has zstd.

The more I think about this, the less I think page compression in redb is a good idea. Here's the latest problem I thought of. redb needs to balance its btree, and it does that based on page fill, so it splits nodes in the btree once their backing page has become full. However, with page compression that heuristic becomes very complex. Either the data needs to be compressed to check its size before determining when to split, or pages will end up with lots of wasted space because nodes will be split and then the data compressed leading to a mostly empty page.

I think compressing the individual values might work okay for you, if you pretrain a shared dictionary and then use that to compress/decompress each value. It's a little annoying since you'll need to collect a representative sample of values to train that dictionary, but in the end you may get even better compression.

**Comment by vlmutolo** [comment]
> Another thing you could try is compression at the file system level. btrfs supports transparent compression and has zstd
I tried this, but I didn't love the requirement that any system running this software would have to use btrfs. There are also compressed FUSE filesystems, but it's still an annoying requirement for the host system.

> I think compressing the individual values might work okay for you, if you pretrain a shared dictionary and then use that to compress/decompress each value. 
That's an excellent point about the shared dictionary. I'll have to give that a shot. 

>  redb needs to balance its btree, and it does that based on page fill, so it splits nodes in the btree once their backing page has become full. However, with page compression that heuristic becomes very complex
Yeah I can see how it would be tough to retrofit compression onto that kind of btree design.

**Comment by cberner** [comment]
Going to go ahead and close this, since it doesn't seem like a good fit for redb's design. Feel free to re-open if you come up with an elegant way to implement it though!
